### PR TITLE
Team: draw multiple shapes

### PR DESCRIPTION
**Pull request by em-henken**

### CODE REVIEW

**Review by thehabes — DISMISSED** [review]
I went through the normal "pick a point flow" and chose a point in the Midwest, United States.

I then went to the Annotation viewer to see the result.

https://map.rerum.io/annotation-viewer.html?data=https://store.rerum.io/v1/id/6578a46135521387cec84c90

I did not see my Point.  After looking for it, I found it in an odd place

image

The coordinates you are capturing are inverted.  It is valid GeoJSON, but GeoJSON states that the order of the point pairs captured in the `coordinates` property are to be in Longitude, Latitude order.  Yours are in Latitude, Longitude order and so do not appear in the correct place on the web map.

When choosing to make a Polygon, when I click for my first point in the Polygon it zooms in.  It should not zoom in when users are trying to draw a Polygon.

As I add more points, they do not appear as a connected shape

image

but as you make the shape it should appear to have an area (like this from geojson.io)

image

After I save the Polygon and go to view it in the Annotation viewer, the Polygon does not load.  

https://map.rerum.io/annotation-viewer.html?data=https://store.rerum.io/v1/id/6578a662e4b2e32bee881f3b

While processing the Annotation, it throws an error
```
Uncaught (in promise) Error: Invalid LatLng object: (undefined, undefined)
    at new v (LatLng.js:32:9)
    at we (GeoJSON.js:236:9)
    at Pe (GeoJSON.js:249:24)
    at Pe (GeoJSON.js:248:4)
    at ye (GeoJSON.js:204:13)
    at i.addData (GeoJSON.js:118:15)
    at i.addData (GeoJSON.js:108:11)
    at i.initialize (GeoJSON.js:93:9)
    at new i (Class.js:24:20)
    at Object.Me (GeoJSON.js:437:9)
```
This means it is not valid GeoJSON in this case.

Please look at these errors and decide how to move forward.  We cannot move this into main until all the base functionality is successful.

**Comment by tobradovic** [comment]
Examples of new edits:
Point:
image

LineString: 
image

Polygon:
image

**Review by yashaswini-slu — APPROVED** [review]
As a user at some point of time, i have come out of drawing the shape, its not happening
Screenshot 2024-02-10 at 12 11 20

**Review by yashaswini-slu — CHANGES_REQUESTED** [review]
The double click is not working has expected
Screenshot 2024-02-16 at 19 57 35

**Review by yashaswini-slu — DISMISSED** [review]
@tobradovic the new design looks good different colors while drawing.
but i still see the issue, there is no end to stop drawing, when we double click the color of the image changes
Screenshot 2024-03-27 at 14 34 30

but i can continue drawing the shape 
Screenshot 2024-03-27 at 14 34 40

**Comment by tobradovic** [comment]
Note on previous commit:

The functions and boundaries that I decided on in the `isCloseToPoint` function are based solely on a few rounds of trial and error.

**Review by yashaswini-slu — CHANGES_REQUESTED** [review]
the double click functionality is good, but there is a issue, the coordinates are getting added multiple times
Screenshot 2024-04-01 at 09 49 27
Screenshot 2024-04-01 at 09 49 48

there are 5 points, but on the JSON we can see the latitude and longitude being repeated, only one should be repeated which represent the end of the shape.

**Comment by tobradovic** [comment]
Note on previous commit:

Problem: the PreviewMap was adding points twice because the 'click' listener was activated twice.
Solution: In the 'dblclick' listener, pop off the extra lat and longs that are unneeded.

**Comment by tobradovic** [comment]
New UI with messages about button functionality:

image

**Review by yashaswini-slu — CHANGES_REQUESTED** [review]
Screenshot 2024-04-07 at 18 59 11
there is repetition for each shape, can we make it common for all, and its taking space.

**Comment by tobradovic** [comment]
New text:

image

**Review by yashaswini-slu — CHANGES_REQUESTED** [review]
Screenshot 2024-04-08 at 13 38 41
we can show this has a tooltip for each icons, it will look good.

File: public/components/layout.js [hunk 1]
```diff
@@ -219,18 +219,19 @@ class PointPicker extends HTMLElement {
...
             <div class="card-body">
-                <div>
-                    <label>Latitude</label>
-                    <input id="leafLat" step=".000000001" type="number" />
-                    <label>Longitude</label>
-                    <input id="leafLong" step=".000000001" type="number" />
+                <div class="row">
+                    <button class="geometry-type-button" style="background-image: url('https://www.svgrepo.com/show/532539/location-pin.svg?size=24') "id="pointBtn" title="Click on the desired location to place a point marker."> </button>
+                    <button class="geometry-type-button" style="background-image: url('https://www.svgrepo.com/show/399231/polyline-pt.svg?size=24');" id="polylineBtn" title="Click to add vertices. Can click the first vertex to finish the shape."> </button>
+                    <button class="geometry-type-button" style="background-image: url('https://www.svgrepo.com/show/399227/polygon-pt.svg?size=24');" id="polygonBtn" title="Clear the map of all selections. Alternatively, click any of the geometry type buttons."> </button>
```
Comment (thehabes) [inline public/components/layout.js:230]: I bet this is not the `title` you want for the polygon button.

image

File: public/components/layout.js [hunk 2]
```diff
@@ -239,64 +240,222 @@ class PointPicker extends HTMLElement {
...
+    pointList = [];
+    marker;
+    markerGroup;
+    previewMap;
+    isCompleteShape = false
+
+    leafletIcon = L.icon({
+        iconUrl: 'marker-icon.png',
+        iconSize: [10, 10],
+    });
```
Comment (thehabes) [inline public/components/layout.js:252]: Remove superfluous `;`

File: public/components/layout.js [hunk 3]
```diff
@@ -239,64 +240,222 @@ class PointPicker extends HTMLElement {
...
         }
-        else {
-            alert("Supply both a latitude and a longitude")
-            return false
+        switch(geometry_type){
+            case 'Point':
+                geo.coordinates = geo.coordinates[0]
+                break
+            case 'LineString':
+                geo.coordinates = geo.coordinates
```
Comment (thehabes) [inline public/components/layout.js:446]: You already know this does nothing...maybe just `break`

File: public/components/layout.js [hunk 4]
```diff
@@ -239,64 +240,222 @@ class PointPicker extends HTMLElement {
...
+                break
+            case 'Polygon':
```
Comment (thehabes) [inline public/components/layout.js:448]: Clean this up a bit

Here is a one liner that accomplishes the same thing (I think, just typed it out I did not test it)
```javascript
geo.coordinates = [...geo.coordinates, ...[geo.coordinates[0]]]
```

**Review by thehabes — CHANGES_REQUESTED** [review]
Confirmed that it is functional and is generating valid Features and a valid Feature Collection!  I believe this put you through an authentic professional task, and one professionals would readily label as "increased difficulty". 

I appreciate the documentation, it is very well done.

See inline comments for a couple touch ups.  Make sure to go through the code and remove all instances of superfluous `,` and `;`

File: public/components/layout.js (hunk 4 above)
Comment (tobradovic) [reply public/components/layout.js:447]: The requested changes have been added in new commits

**Comment by tobradovic** [comment]
New text format: tooltips when you hover over the geometry buttons

image
image
image
image

**Review by yashaswini-slu — APPROVED** [review]

**Review by thehabes — CHANGES_REQUESTED** [review]
Still some dangling semicolons but that's the only thing I noticed.  Make sure your code does not introduce unnecessary semicolons.  I know it's annoying and a small thing, but  Javascript has evolved to not require those to end a line while also using them in specialized syntax.  Leaving extra ones around may cause problems that are very hard to find later, because all they are is `;`

Once you clear those out, you can merge this.  The code changes were successful and everything is in order.

**Review by thehabes — APPROVED** [review]
Thanks for the cleanup!  If you have done enough testing and are confident it is ready to go forward push the button @yashaswini-slu.  

Good work everyone.

**Pull request merged**